src/serializer/ResidualFunctions.js

```diff
@@ -249,49 +255,69 @@ export class ResidualFunctions {
           invariant(t.isCallExpression(funcNode)); // .bind call
           body = getFunctionBody(instance);
         }
+        let pushToBody = elem => {
+          if (instance.declarationBodyOverride) instance.declarationBodyOverride.unshift(elem);
+          else body.push(elem);
+        };
         let declaration = t.variableDeclaration("var", [t.variableDeclarator(funcId, funcNode)]);
-        body.push(declaration);
+        pushToBody(declaration);
         let prototypeId = this.functionPrototypes.get(functionValue);
         if (prototypeId !== undefined) {
           let id = this.locationService.getLocation(functionValue);
           invariant(id !== undefined);
-          body.push(
+          pushToBody(
             t.variableDeclaration("var", [
               t.variableDeclarator(prototypeId, t.memberExpression(id, t.identifier("prototype"))),
             ])
           );
         }
       };
+      // Split instances into normal or additional functions (whose bodies have been rewritten)
+      let normalInstances = [];
+      let additionalFunctionInstances = [];
+      for (let instance of instances) {
+        if (rewrittenAdditionalFunctions.has(instance.functionValue)) additionalFunctionInstances.push(instance);
+        else normalInstances.push(instance);
+      }
+
+      let rewrittenBody = rewrittenAdditionalFunctions.get(instances[0].functionValue);
 
-      if (shouldInline || instances.length === 1 || usesArguments) {
-        this.statistics.functionClones += instances.length - 1;
+      // Creates a new function for each instance.
+      let naiveProcessInstances = (funcInstances, functionBody, isRewrittenFunction) => {
+        this.statistics.functionClones += funcInstances.length - 1;
 
-        for (let instance of instances) {
+        for (let instance of funcInstances) {
           let { functionValue, serializedBindings, scopeInstances } = instance;
           let id = this.locationService.getLocation(functionValue);
           invariant(id !== undefined);
           let funcParams = params.slice();
           let funcNode = t.functionExpression(
             null,
             funcParams,
-            ((t.cloneDeep(funcBody): any): BabelNodeBlockStatement)
+            // We can't clone the body of rewritten functions because we may need to
+            // modify it with additional function declarations.
```

Comment:
At this point I'm lost.

Reply:
I also don't understand the comment.

Reply:
We shouldn't clone the body of a rewritten function because we might need to add function declarations to that body for functions defined inside of that additional function (see line 258). Additionally we shouldn't be running the closure ref fixup because the additional function's body is post-serialization so the identifiers are all correct (this was the original reason why we cloned the bodies).

I'll change the wording of the comment.